plans/benchmarks/benchmarks.go

```diff
@@ -164,3 +172,101 @@ func BarrierBench(runenv *runtime.RunEnv) error {
 
 	return nil
 }
+
+func SubtreePublishBench(runenv *runtime.RunEnv) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
+	defer cancel()
+	watcher, writer := sync.MustWatcherWriter(ctx, runenv)
+	defer watcher.Close()
+	defer writer.Close()
+
+	if err := sync.WaitNetworkInitialized(ctx, runenv, watcher); err != nil {
+		return err
+	}
+
+	var msg []byte
+	st := sync.Subtree{
+		GroupKey:    "messages",
+		PayloadType: reflect.TypeOf(&msg),
+		KeyFunc: func(val interface{}) string {
+			return string(*val.(*[]byte))
+		}}
+
+	seq, err := writer.Write(ctx, &st, runenv.TestRun)
+	if err != nil {
+		return err
+	}
+
+	// Create several tests up to 1MB
+	// Anything over 1500 is likely to have ethernet fragmentation
+	// Do these sizes make any sense at all? I really don't know.
+	var tests []*subtreeBenchCfg
+	for size := 64; size <= 1024*1024; size = size << 1 {
+		buf := make([]byte, size)
+		name := fmt.Sprintf("subtree_bench_%s_bytes", string(size))
+		t := subtreeBenchCfg{
+			Name:  name,
+			Gauge: runtime.NewGauge(runenv, name, fmt.Sprintf("Time to transmmit %d bytes", size)),
+			Data:  buf,
+		}
+		tests = append(tests, &t)
+
+	}
+
+	if seq == 1 {
+		// Publish information to the subtree and wait for all to receive it.
+		return subtreePublisher(ctx, runenv, writer, watcher, &st, tests)
+	} else {
+		// Subscribe to the subtree as quickly as quickly as possible.
+		return subtreeSubscriber(ctx, runenv, writer, watcher, &st, tests)
+	}
+}
+
+func subtreePublisher(ctx context.Context, runenv *runtime.RunEnv, writer *sync.Writer, watcher *sync.Watcher, st *sync.Subtree, tests []*subtreeBenchCfg) error {
+
+	rand.Seed(time.Now().UnixNano())
+
+	// The sender doesn't publish metrics.
+	// Just keep everyone on the same page, and publsh fresh data.
+	// Wait until all are ready,
+	// publish
+	// begin the test
+	// Repeat.
+
+	for i := 0; i <= 100000; i++ {
+		for _, tst := range tests {
+			readyState := sync.State(fmt.Sprintf("ready_%d_%s", i, tst.Name))
+			testState := sync.State(fmt.Sprintf("test_%d_%s", i, tst.Name))
+			rand.Read(tst.Data)
+			_, err := writer.SignalEntry(ctx, readyState)
+			if err != nil {
+				return err
+			}
+			<-watcher.Barrier(ctx, readyState, int64(runenv.TestInstanceCount))
+			writer.SignalEntry(ctx, testState)
```

Comment:
I'm not sure we need any barrier-based synchronisation here. Both publisher and subscriber know how many messages (iterations) we're going to send/receive. We should run them sequentially:

1. First the publisher. It writes each message and records the time it took to do so. I would not usage a Prometheus Gauge, but a Histogram or a Summary. A Gauge is going to give us insufficient data IMO.
2. After the publisher is done, we run the subscribers, who subscribe, start receiving, and record on a histogram the time it took between the last receive and the current one.

While it would be great to run concurrent publishers and subscribers, the resulting scenario would be highly racy and volatile, dependent on timing.